### PR TITLE
Cleanup dispatchers and network transports.

### PR DESCRIPTION
**Pull request by cmelchior**
Closes https://github.com/realm/realm-kotlin/issues/818
Closes https://github.com/realm/realm-kotlin/issues/733

This PR reworks how we track and create dispatchers and network transports, so they are now owned by either each `App` or `Realm` instance. This makes it possible for these instances to close them, when they themselves are closed, thus preventing us from leaking thread pools.

This is made a little bit complicated due to two reasons: 1) The generic CoroutineDispatcher API does not have a close method and 2) We only want to automatically close thread pools and dispatchers controlled by us, not those provided externally.

For this reason, I have introduced two new concepts: `CoroutineDispatcherFactory` and `ManagedCoroutineDispatcher`, both found here: https://github.com/realm/realm-kotlin/pull/1042/files#diff-62a0fa6bf1e3a7353b7796d7e8481de740581e30b5bf92186fad7f2dbdcad2af

The factory is replacing dispatchers in our configuration objects so now those objects no longer hold any references to our own dispatchers. Dispatchers will only be created when asked to, which means the App/Realm can do it. 

The `ManagedCoroutineDispatcher` is then my way of making the close method more easily accessible. I made it a wrapper class with a `get()` method, to access the underlying dispatcher because the CoroutineDispatcher API is an abstract class and our wrapper should only delegate calls. If anyone has a better idea on how to abstract this cleaner I am all ears.

Also, the Realm Finalizer thread was not marked as a daemon thread, preventing the JVM from exiting when it was running. Since there is no way to know when it is safe to close this. I marked it as a daemon thread instead. This might impact how often the cleanup runs, but I'm not 100% sure how to evaluate if it is a problem or not. So for now, I'm relying on the test suite to surface any problems.

Writing tests for this is rather hard, so I only added JVM tests but the assumption is that it will work across all platforms.

**TODO**
- [x] It looks like we still have one rogue thread from synced Realms. Probably the Sync Client thread. The lifecycle of the sync client should probably follow the app instance and needs to be investigated. https://github.com/realm/realm-core/issues/5966 is tracking this.

### CODE REVIEW

**Comment by cmelchior** [comment]
Internal note. After making sure everything is cleaned up, the rogue thread still seem to active (Thread-4):

```
Killing app
2022-10-21 16:37:25.091 30104-30137 REALM                   io.realm.sync.testapp.test           E  Stopping SyncManager
2022-10-21 16:37:25.091 30104-30137 REALM                   io.realm.sync.testapp.test           E  Stopping SyncClient
2022-10-21 16:37:25.091 30104-30137 REALM                   io.realm.sync.testapp.test           E  Stopping sync_client.hpp
2022-10-21 16:37:25.091 30104-30137 REALM                   io.realm.sync.testapp.test           E  Joining worker sync_client.hpp
2022-10-21 16:37:25.091 30104-30137 REALM                   io.realm.sync.testapp.test           E  Joined worker sync_client.hpp
2022-10-21 16:37:25.091 30104-30137 REALM                   io.realm.sync.testapp.test           E  Stopped SyncClient
2022-10-21 16:37:25.091 30104-30137 REALM                   io.realm.sync.testapp.test           E  Stopping sync_client.hpp
2022-10-21 16:37:28.886   376-403   memtrack                system_process                       E  Couldn't load memtrack module
2022-10-21 16:37:28.886   376-403   android.os.Debug        system_process                       W  failed to get memory consumption info: -1
2022-10-21 16:37:33.051   177-766   netd                    netd                                 I  trafficSwapActiveStatsMap() <15.04ms>
2022-10-21 16:37:33.055   177-766   netd                    netd                                 I  tetherGetStats() <2.54ms>
2022-10-21 16:37:35.216 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 13354(523KB) AllocSpace objects, 2(40KB) LOS objects, 49% free, 2956KB/5912KB, paused 16us total 9.070ms
2022-10-21 16:37:35.339 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 28(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 33us total 17.745ms
2022-10-21 16:37:35.463 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 50us total 21.729ms
2022-10-21 16:37:35.588 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 29us total 19.786ms
2022-10-21 16:37:35.709 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 35us total 17.800ms
2022-10-21 16:37:35.822 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 29us total 10.417ms
2022-10-21 16:37:35.939 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 18us total 9.531ms
2022-10-21 16:37:36.061 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 23us total 20.433ms
2022-10-21 16:37:36.168 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 7us total 4.029ms
2022-10-21 16:37:36.274 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 7us total 3.590ms
2022-10-21 16:37:36.379 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 7us total 3.718ms
2022-10-21 16:37:36.487 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 10us total 5.162ms
2022-10-21 16:37:36.596 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 15us total 8.164ms
2022-10-21 16:37:36.712 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 20us total 15.437ms
2022-10-21 16:37:36.831 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 23us total 15.635ms
2022-10-21 16:37:36.951 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 24us total 17.516ms
2022-10-21 16:37:37.065 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 18us total 9.221ms
2022-10-21 16:37:37.183 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 31us total 11.905ms
2022-10-21 16:37:37.304 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 29us total 13.203ms
2022-10-21 16:37:37.427 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 20us total 19.574ms
2022-10-21 16:37:37.544 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 30us total 14.191ms
2022-10-21 16:37:37.662 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 29us total 16.050ms
2022-10-21 16:37:37.785 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 34us total 15.240ms
2022-10-21 16:37:37.892 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 7us total 3.853ms
2022-10-21 16:37:37.997 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 7us total 4.224ms
2022-10-21 16:37:38.103 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 7us total 3.852ms
2022-10-21 16:37:38.208 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 7us total 3.822ms
2022-10-21 16:37:38.315 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 11us total 4.976ms
2022-10-21 16:37:38.426 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 15us total 8.373ms
2022-10-21 16:37:38.539 30104-30137 nc.testapp.tes          io.realm.sync.testapp.test           I  Explicit concurrent copying GC freed 6(16KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2955KB/5911KB, paused 13us total 5.585ms
2022-10-21 16:37:43.540 30104-30137 System.out              io.realm.sync.testapp.test           I  Remaining threads:
2022-10-21 16:37:43.541 30104-30137 System.out              io.realm.sync.testapp.test           I  --------------------------------
2022-10-21 16:37:43.542 30104-30137 System.out              io.realm.sync.testapp.test           I  DefaultDispatcher-worker-1 - Is Daemon true - Is Alive true
2022-10-21 16:37:43.542 30104-30137 System.out              io.realm.sync.testapp.test           I  Binder:30104_1 - Is Daemon false - Is Alive true
2022-10-21 16:37:43.542 30104-30137 System.out              io.realm.sync.testapp.test           I  Jit thread pool worker thread 0 - Is Daemon true - Is Alive true
2022-10-21 16:37:43.542 30104-30137 System.out              io.realm.sync.testapp.test           I  HeapTaskDaemon - Is Daemon true - Is Alive true
2022-10-21 16:37:43.542 30104-30137 System.out              io.realm.sync.testapp.test           I  ReferenceQueueDaemon - Is Daemon true - Is Alive true
2022-10-21 16:37:43.542 30104-30137 System.out              io.realm.sync.testapp.test           I  Binder:30104_3 - Is Daemon false - Is Alive true
2022-10-21 16:37:43.542 30104-30137 System.out              io.realm.sync.testapp.test           I  InstrumentationConnectionThread - Is Daemon false - Is Alive true
2022-10-21 16:37:43.542 30104-30137 System.out              io.realm.sync.testapp.test           I  Profile Saver - Is Daemon true - Is Alive true
2022-10-21 16:37:43.542 30104-30137 System.out              io.realm.sync.testapp.test           I  Instr: androidx.test.runner.AndroidJUnitRunner - Is Daemon false - Is Alive true
2022-10-21 16:37:43.542 30104-30137 System.out              io.realm.sync.testapp.test           I  DefaultDispatcher-worker-2 - Is Daemon true - Is Alive true
2022-10-21 16:37:43.543 30104-30137 System.out              io.realm.sync.testapp.test           I  RealmFinalizingDaemon - Is Daemon true - Is Alive true
2022-10-21 16:37:43.543 30104-30137 System.out              io.realm.sync.testapp.test           I  DefaultDispatcher-worker-3 - Is Daemon true - Is Alive true
2022-10-21 16:37:43.543 30104-30137 System.out              io.realm.sync.testapp.test           I  Binder:30104_2 - Is Daemon false - Is Alive true
2022-10-21 16:37:43.543 30104-30137 System.out              io.realm.sync.testapp.test           I  FinalizerWatchdogDaemon - Is Daemon true - Is Alive true
2022-10-21 16:37:43.543 30104-30137 System.out              io.realm.sync.testapp.test           I  main - Is Daemon false - Is Alive true
2022-10-21 16:37:43.543 30104-30137 System.out              io.realm.sync.testapp.test           I  Signal Catcher - Is Daemon true - Is Alive true
2022-10-21 16:37:43.543 30104-30137 System.out              io.realm.sync.testapp.test           I  FinalizerDaemon - Is Daemon true - Is Alive true
2022-10-21 16:37:43.543 30104-30137 System.out              io.realm.sync.testapp.test           I  Thread-4 - Is Daemon false - Is Alive true
2022-10-21 16:37:43.543 30104-30137 System.out              io.realm.sync.testapp.test           I  Total threads: [Thread[DefaultDispatcher-worker-1,5,main], Thread[Binder:30104_1,5,main], Thread[Jit thread pool worker thread 0,5,main], Thread[HeapTaskDaemon,5,system], Thread[ReferenceQueueDaemon,5,system], Thread[Binder:30104_3,5,main], Thread[InstrumentationConnectionThread,5,main], Thread[Profile Saver,5,system], Thread[Instr: androidx.test.runner.AndroidJUnitRunner,5,main], Thread[DefaultDispatcher-worker-2,5,main], Thread[RealmFinalizingDaemon,5,main], Thread[DefaultDispatcher-worker-3,5,main], Thread[Binder:30104_2,5,main], Thread[FinalizerWatchdogDaemon,5,system], Thread[main,5,main], Thread[Signal Catcher,5,system], Thread[FinalizerDaemon,5,system], Thread[Thread-4,10,main]]
2022-10-21 16:37:43.544 30104-30137 System.out              io.realm.sync.testapp.test           I  Active threads: 12
```

Maybe the thread is not a sync thread after all? 🤔

**Comment by cmelchior** [comment]
Root cause for the Sync Client thread issue is this one: https://github.com/realm/realm-core/issues/5966

We need new C-API functionality before being able to clean it up.

**Comment by cmelchior** [comment]
This should now be ready for review. Note, we are still leaking the Sync Client thread, but there is a core issue tracking it so I expect we will be able to fix that shortly.

File: packages/cinterop/src/jvm/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt
Comment (rorbech) [inline packages/cinterop/src/jvm/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt:1187]: We have previously tried to put all configuration in the library to ensure that we have the exact same configuration on all platforms. I guess it is quite tedious with having to wrap the enums, so wouldn't block approval, but could consider to do a no-arg-method that set it to `IMMEDIATELY` ... but I guess that could also be implemented incorrectly on each platform 🤪  ... but as a reviewer it would be easier to spot errors in that, than spotting that you have forgotten the same configuration call on each platform.

**Review by rorbech — COMMENTED** [review]
Awesome. I think the concept is right, just a bit confused about the implementation. I. would somehow collapse the factory and managed dispatcher and/or simplify it with delegation or something like that. I have tried to put some hints, but ping me if you don't get it. Might be that I have just been tricked by using `dispatcher` variable names for the factories too.

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/util/CoroutineDispatcherFactory.kt
Comment (cmelchior) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/util/CoroutineDispatcherFactory.kt:20]: Not really. I just wanted to future-proof the API since we have constructs in our code for creating multithreaded dispatchers. But I can remove it again if it is confusing.

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/util/CoroutineDispatcherFactory.kt
Comment (cmelchior) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/util/CoroutineDispatcherFactory.kt:46]: Ah crap, delegating only work on interfaces 😭

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/util/CoroutineDispatcherFactory.kt
Comment (cmelchior) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/util/CoroutineDispatcherFactory.kt:14]: We could also do that, I guess it feels slightly weird that the manager can close references it handed out without the dispatcher being involved, but at the same, it would help to get rid of the ManagedCoroutineDispatcher.

So I could probably agree that since both approaches have some annoyance, going with the one that has the least amount of code is nicer.

Comment (cmelchior) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/util/CoroutineDispatcherFactory.kt:14]: Hmm, actually that is not going to work I think. the problem is that the `RealmConfiguration` can be used to open multiple Realm instances, all with their own dispatchers, so nothing inside the configuration is allowed to track handed-out configurations.

We would then end up with `RealmCoroutineFactoryFactory` ... which seems even more silly 🙈

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/util/CoroutineDispatcherFactory.kt
Comment (clementetb) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/util/CoroutineDispatcherFactory.kt:46]: You could even use inline classes and polymorfism to reduce the complexity, see: https://gist.github.com/clementetb/58b592ee766b440d95f0a67b60b7e8ad

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/RealmConfiguration.kt
Comment (clementetb) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/RealmConfiguration.kt:131]: I think you could remove the null assertion and delete the if with:

`notificationDispatcher?.let { CoroutineDispatcherFactory.external(it) }`

and

`writeDispatcher?.let { CoroutineDispatcherFactory.external(it) }`

then make make the dispatchers constructor parameters in `RealmConfigurationImpl` nullable, to finally 
change the parameters in the `ConfigurationImpl` constructor to

`notificationDispatcher ?: CoroutineDispatcherFactory.internal(name)` 

🤯

File: packages/cinterop/src/jvm/kotlin/io/realm/kotlin/internal/interop/gc/NativeContext.kt
Comment (clementetb) [inline packages/cinterop/src/jvm/kotlin/io/realm/kotlin/internal/interop/gc/NativeContext.kt:31]: Dont we manually setup a priority? as we do in `CoroutineUtilsSharedJvm`? If not set it will default to the priority of the thread it was created from:

>When code running in some thread creates a new Thread object, the new thread has its priority initially set equal to the priority of the creating thread, and is a daemon thread if and only if the creating thread is a daemon.

We might like to have some control on it.

File: packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/internal/SubscriptionSetImpl.kt
Comment (clementetb) [inline packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/internal/SubscriptionSetImpl.kt:87]: How can we make it less dangerous?

File: packages/test-sync/src/jvmTest/kotlin/io/realm/kotlin/jvm/RealmTests.kt
Comment (clementetb) [inline packages/test-sync/src/jvmTest/kotlin/io/realm/kotlin/jvm/RealmTests.kt:1]: Missing copyright

**Review by clementetb — APPROVED** [review]
Looks great to me, I added some minor comments.

File: packages/cinterop/src/jvm/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt
Comment (cmelchior) [reply packages/cinterop/src/jvm/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt:1187]: Yes, that is the default, but if you are offline it will keep running indefinitely preventing the Sync thread from exiting, and thus the JVM.

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/util/CoroutineDispatcherFactory.kt
Comment (cmelchior) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/util/CoroutineDispatcherFactory.kt:46]: Yeah, that does improve it 👍

File: packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/internal/SubscriptionSetImpl.kt
Comment (cmelchior) [reply packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/internal/SubscriptionSetImpl.kt:87]: This should be safe until we introduce a public DynamicRealm as we only expose subscriptions through the Realm class. I expanded this comment a bit to make it more clear.

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/RealmConfiguration.kt
Comment (cmelchior) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/RealmConfiguration.kt:131]: True, but that would also split the logic where we set up the RealmConfiguration. I agree the above code is a bit messy. I have tried to move it around a little to hopefully make it more clear.

Comment (clementetb) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/RealmConfiguration.kt]: Yes, I came to say that you could save the `let` by setting the value in the builder.

File: packages/test-sync/src/jvmTest/kotlin/io/realm/kotlin/jvm/RealmTests.kt
Comment (rorbech) [inline packages/test-sync/src/jvmTest/kotlin/io/realm/kotlin/jvm/RealmTests.kt:40]: Shouldn't we have a similar test on native?

Comment (cmelchior) [reply packages/test-sync/src/jvmTest/kotlin/io/realm/kotlin/jvm/RealmTests.kt:40]: Ideally yes, but I don't think we have the same capabilities on Native to introspect the state of threads, so I assumed that for the things we are interested in, testing on JVM should be sufficient?

Comment (rorbech) [reply packages/test-sync/src/jvmTest/kotlin/io/realm/kotlin/jvm/RealmTests.kt:40]: Yeah, I guess this ensures that we are invoking the expected underlying framework calls, so this do ensure that our overall implementation should be right. But then we should maybe do a test that shows that we don't close custom dispatchers or create new threads if we pass them in?

Comment (cmelchior) [reply packages/test-sync/src/jvmTest/kotlin/io/realm/kotlin/jvm/RealmTests.kt:40]: Good point. That is currently missing. I will add that 👍

File: packages/test-base/src/jvmTest/kotlin/io/realm/kotlin/test/jvm/RealmTests.kt
Comment (nhachicha) [inline packages/test-base/src/jvmTest/kotlin/io/realm/kotlin/test/jvm/RealmTests.kt:45]: Be careful this only returns the number of threads within the current `ThreadGroup`

**Review by nhachicha — COMMENTED** [review]
Nice refactor 🙌

File: packages/cinterop/src/jvm/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt
Comment (cmelchior) [reply packages/cinterop/src/jvm/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt:793]: Good point. It will require some changes from https://github.com/realm/realm-kotlin/pull/1092 though. So we should probably not merge this PR until that one has been merged first.

**Review by rorbech — APPROVED** [review]
:shipit:

**Pull request merged**
